lib/broadway/producer.ex

```diff
@@ -236,4 +303,129 @@ defmodule Broadway.Producer do
+        # we close the rate limiting and don't emit anything.
+        allowed when allowed <= 0 ->
+          {%{rate_limiting | state: :closed}, []}
+
+        allowed ->
+          {allowed_left, probably_emittable, buffer} = dequeue_many(buffer, allowed, [])
+
+          {rate_limiting_state, messages_to_emit, messages_to_buffer} =
+            rate_limit_messages(
+              table_name,
+              probably_emittable,
+              _probably_emittable_count = allowed - allowed_left
+            )
+
+          new_buffer = enqueue_batch_r(buffer, messages_to_buffer)
+
+          rate_limiting = %{
+            rate_limiting
+            | message_buffer: new_buffer,
+              state: rate_limiting_state
+          }
+
+          if draining? and :queue.is_empty(new_buffer) do
+            cancel_consumers(state)
+          end
+
+          {rate_limiting, messages_to_emit}
+      end
+
+    {%{state | rate_limiting: rate_limiting}, messages_to_emit}
+  end
+
+  defp reverse_split_demand(rest, 0, acc) do
+    {0, acc, rest}
+  end
+
+  defp reverse_split_demand([], demand, acc) do
+    {demand, acc, []}
+  end
+
+  defp reverse_split_demand([head | tail], demand, acc) do
+    reverse_split_demand(tail, demand - 1, [head | acc])
+  end
+
+  defp dequeue_many(queue, demand, acc) do
+    case :queue.out(queue) do
+      {{:value, list}, queue} ->
+        case reverse_split_demand(list, demand, acc) do
+          {0, acc, []} ->
+            {0, Enum.reverse(acc), queue}
+
+          {0, acc, rest} ->
+            {0, Enum.reverse(acc), :queue.in_r(rest, queue)}
+
+          {demand, acc, []} ->
+            dequeue_many(queue, demand, acc)
+        end
+
+      {:empty, queue} ->
+        {demand, Enum.reverse(acc), queue}
+    end
+  end
+
+  defp enqueue_batch(queue, _list = []), do: queue
+  defp enqueue_batch(queue, list), do: :queue.in(list, queue)
+
+  defp enqueue_batch_r(queue, _list = []), do: queue
+  defp enqueue_batch_r(queue, list), do: :queue.in_r(list, queue)
+
+  defp rate_limit_messages(_state, [], _count) do
+    {[], []}
+  end
+
+  defp rate_limit_messages(table_name, messages, message_count) do
+    case RateLimiter.rate_limit(table_name, message_count) do
+      # If no more messages are allowed, we're rate limited but we're able
+      # to emit all messages that we have.
+      0 ->
+        {:closed, messages, _to_buffer = []}
+
+      # We were able to emit all messages and still more messages are allowed,
+      # so the rate limiting is "open".
+      left when left > 0 ->
+        {:open, messages, _to_buffer = []}
+
+      # We went over the rate limit, so we split (on negative index) the messages
+      # we were able to emit and close the rate limiting.
+      overflow when overflow < 0 ->
+        {emittable, to_buffer} = Enum.split(messages, overflow)
+        {:closed, emittable, to_buffer}
+    end
+  end
+
+  defp schedule_next_handle_demand_if_rate_limiting_open(state) do
```

Comment:
Extract the `rate_limiting` key on the function clause to avoid multiple `state.rate_limiting` calls.